master/static/migrations/20230405164440_add-summary-metrics.tx.up.sql

```diff
@@ -0,0 +1,272 @@
+-- We need to declare start_timestamp before this migration starts
+-- to avoid missing metrics reported while this migration runs in background cases.
+DO $$ DECLARE start_timestamp timestamptz := NOW(); BEGIN
+
+ALTER TABLE trials
+    ADD COLUMN IF NOT EXISTS summary_metrics jsonb NOT NULL DEFAULT '{}',
+    ADD COLUMN IF NOT EXISTS summary_metrics_timestamp timestamptz;
+
+-- Invalidate summary_metrics_timestamp for trials that have a metric added since.
+WITH max_training as (
+     SELECT trial_id, max(steps.end_time) as last_reported_metric FROM steps
+     JOIN trials ON trials.id = trial_id
+     WHERE summary_metrics_timestamp IS NOT NULL
+     GROUP BY trial_id
+)
+UPDATE trials SET summary_metrics_timestamp = NULL FROM max_training WHERE
+    max_training.trial_id = trials.id AND
+    summary_metrics_timestamp IS NOT NULL AND
+    last_reported_metric > summary_metrics_timestamp;
+
+WITH max_validation as (
+     SELECT trial_id, max(validations.end_time) as last_reported_metric FROM validations
+     JOIN trials ON trials.id = trial_id
+     WHERE summary_metrics_timestamp IS NOT NULL
+     GROUP BY trial_id
+)
+UPDATE trials SET summary_metrics_timestamp = NULL FROM max_validation WHERE
+     max_validation.trial_id = trials.id AND
+     summary_metrics_timestamp IS NOT NULL AND
+     last_reported_metric > summary_metrics_timestamp;
+
+-- Returns pairs of metric names and trial_ids and if they are numeric or not.
+WITH training_trial_metrics as (
+SELECT
+    name,
+    trial_id,
+    sum(entries) FILTER (WHERE metric_type != 'number') as nonumbers
+FROM (
+    SELECT
+    name,
+    CASE
+        WHEN (metrics->'avg_metrics'->name)::text = '"Infinity"'::text THEN 'number'
+        WHEN (metrics->'avg_metrics'->name)::text = '"-Infinity"'::text THEN 'number'
+        WHEN (metrics->'avg_metrics'->name)::text = '"NaN"'::text THEN 'number'
+        ELSE jsonb_typeof(metrics->'avg_metrics'->name)
```

Comment:
came up in testing top trials by metric

old code would sometimes let people get away with report the string ```"1.0"``` and we would cast it to the float ```1.0``` in some queries.

should this migration and ingestion code treat ```"1.0"``` as ```1.0```?

Reply:
no, I don't think this is a kind of behavior we should support. that doesn't seem to me like something a user would rely upon.